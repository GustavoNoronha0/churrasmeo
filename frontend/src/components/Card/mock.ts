import { CardProps } from '.'

export const CardMock: CardProps = {
  date: new Date(),
  title: 'Churras meo',
  countParticipants: 10,
  value: 250
}
