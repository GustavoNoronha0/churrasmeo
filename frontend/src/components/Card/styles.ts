import styled, { css } from 'styled-components'
import { BsPeople } from 'react-icons/bs'
import { AiOutlineDollar } from 'react-icons/ai'

import { motion, AnimatePresence } from 'framer-motion'

export const Animate = styled(AnimatePresence)``
export const Motion = styled(motion.div).attrs({
  transition: { ease: 'easeInOut', duration: 0.5 },
  initial: { opacity: 0, scale: 0.6 },
  animate: { opacity: 1, scale: 1 },
  exit: { opacity: 0, scale: 0.5 },
  'data-testid': 'motion'
})`
  position: relative;

  max-width: 80%;
  width: auto;
  margin: auto;
`

export const Wrapper = styled.div.attrs({
  'data-testid': 'WrapperCard'
})`
  width: fit-content;

  display: flex;
  justify-content: center;
  align-items: center;

  top: 5;
  right: 0;
`

export const Card = styled.button`
  ${({ theme }) => css`
    box-shadow: 0 4px 8px rgba(0, 0, 0, 0.2);
    border-radius: ${theme.border.radius};
    display: flex;
    align-items: center;
    cursor: pointer;
    border: none;

    &:hover {
      box-shadow: 1px 1px 1px 1px rgba(0, 0, 0, 0.2);
    }
  `}
`

export const DivCard = styled.div`
  ${({ theme }) => css`
    background-color: ${theme.colors.white};
    box-shadow: 0 4px 8px rgba(0, 0, 0, 0.2);
    border-radius: ${theme.border.radius};
    padding: 20px;
    width: 300px;
    height: 200px;

    display: flex;
    flex-direction: column;
    align-items: left;
  `}
`

export const TitleCard = styled.p`
  ${({ theme }) => css`
    font-style: normal;
    font-weight: bold;
    font-size: 21px;
    line-height: 25px;
    color: ${theme.colors.black};
  `}
`

export const Header = styled.div`
  display: flex;
  flex-direction: column;

  top: 0;
`

export const Date = styled.h2`
  ${({ theme }) => css`
    font-style: normal;
    font-weight: 800;
    font-size: 28px;
    line-height: 33px;
    color: ${theme.colors.baseColorYellow};
  `}
`

export const Content = styled.div`
  display: flex;
  justify-content: space-between;
  position: relative;
  height: 100px;
`

export const CountParticipants = styled.h4`
  display: flex;
  flex-direction: row;
  position: absolute;
  bottom: 0;
`

export const Value = styled.h4`
  display: flex;
  flex-direction: row;
  position: absolute;
  bottom: 0;
  right: 0;
`

export const IconPeople = styled(BsPeople)`
  ${({ theme }) => css`
    font-size: 18px;
    margin-right: 5px;
    color: ${theme.colors.baseColorYellow};
  `}
`

export const IconDollar = styled(AiOutlineDollar)`
  ${({ theme }) => css`
    font-size: 18px;
    margin-right: 5px;
    color: ${theme.colors.baseColorYellow};
  `}
`
