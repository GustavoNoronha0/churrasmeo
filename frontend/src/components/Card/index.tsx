import * as S from './styles'
export * from './mock'
import React from 'react'
import { formatData, sumValues } from '@/utils/helpers'
import { ParticipantsModel } from '@/services/models/participants'

export type TypeCard = 'success' | 'error'

export type CardProps = {
  date: Date
  title: string
  countParticipants: number
  participants: ParticipantsModel[]
  onClick?: (event: React.MouseEvent<HTMLButtonElement>) => void
}

const Card = ({
  date,
  title,
  countParticipants,
  participants,
  onClick
}: CardProps) => {
  const participantsValues = participants.map(
    (participant) => participant.value
  )
  return (
    <S.Wrapper>
      <S.Animate>
        <S.Motion>
          <S.Card onClick={onClick}>
            <S.DivCard>
              <S.Header>
                <S.Date>{formatData(date.toString())}</S.Date>
                <S.TitleCard>{title}</S.TitleCard>
              </S.Header>
              <S.Content>
                <S.CountParticipants>
                  <S.IconPeople /> {countParticipants}
                </S.CountParticipants>
                <S.Value>
                  <S.IconDollar />
                  {sumValues(participantsValues)}
                </S.Value>
              </S.Content>
            </S.DivCard>
          </S.Card>
        </S.Motion>
      </S.Animate>
    </S.Wrapper>
  )
}

export default Card
