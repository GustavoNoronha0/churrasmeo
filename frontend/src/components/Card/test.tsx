import { render, screen } from '@/utils/tests/helpers'

import Card from '.'
import userEvent from '@testing-library/user-event'

describe('<Card />', () => {
  const onClick = jest.fn()
  it('should render a card default', () => {
    const { container } = render(
      <Card
        countParticipants={10}
        date={new Date()}
        title="Titulo da modal"
        participants={[
          {
            id: 'any_id',
            barbecue: { id: 'any_id' },
            name: 'any_name',
            value: 1,
            value_suggestions_with_drink: 1,
            value_suggestions_with_out_drink: 1
          },
          {
            id: 'any_id2',
            barbecue: { id: 'any_id' },
            name: 'any_name',
            value: 1,
            value_suggestions_with_drink: 1,
            value_suggestions_with_out_drink: 1
          }
        ]}
        onClick={onClick}
      />
    )

    expect(screen.getByText('Titulo da modal')).toBeInTheDocument()
    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render a card with color style', () => {
    render(
      <Card
        countParticipants={10}
        date={new Date()}
        title="Titulo da modal"
        participants={[
          {
            id: 'any_id',
            barbecue: { id: 'any_id' },
            name: 'any_name',
            value: 1,
            value_suggestions_with_drink: 1,
            value_suggestions_with_out_drink: 1
          },
          {
            id: 'any_id2',
            barbecue: { id: 'any_id' },
            name: 'any_name',
            value: 1,
            value_suggestions_with_drink: 1,
            value_suggestions_with_out_drink: 1
          }
        ]}
        onClick={onClick}
      />
    )

    expect(screen.getByText('Titulo da modal')).toBeInTheDocument()
    const wrapperCard = screen.getByTestId('WrapperCard')
    expect(wrapperCard).toHaveStyle({
      width: 'fit-content'
    })
  })

  it('should on card click call onClick function', () => {
    const mockCallback = jest.fn((number: number) => number + 1)

    const onCardClick = jest.fn(() => mockCallback(1))

    render(
      <Card
        countParticipants={10}
        date={new Date()}
        title="Titulo da modal"
        participants={[
          {
            id: 'any_id',
            barbecue: { id: 'any_id' },
            name: 'any_name',
            value: 1,
            value_suggestions_with_drink: 1,
            value_suggestions_with_out_drink: 1
          },
          {
            id: 'any_id2',
            barbecue: { id: 'any_id' },
            name: 'any_name',
            value: 1,
            value_suggestions_with_drink: 1,
            value_suggestions_with_out_drink: 1
          }
        ]}
        onClick={onCardClick}
      />
    )

    const cardToClick = screen.getByText('Titulo da modal')

    userEvent.click(cardToClick)

    expect(onCardClick).toHaveBeenCalledTimes(1)
    expect(onCardClick).toReturnWith(2)
  })
})
