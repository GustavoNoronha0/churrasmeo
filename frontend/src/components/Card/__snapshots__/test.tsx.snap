// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card /> should render a card default 1`] = `
.c1 {
  position: relative;
  max-width: 80%;
  width: auto;
  margin: auto;
}

.c0 {
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  top: 5;
  right: 0;
}

.c2 {
  box-shadow: 0 4px 8px rgba(0,0,0,0.2);
  border-radius: 0.4rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  border: none;
}

.c2:hover {
  box-shadow: 1px 1px 1px 1px rgba(0,0,0,0.2);
}

.c3 {
  background-color: #FAFAFA;
  box-shadow: 0 4px 8px rgba(0,0,0,0.2);
  border-radius: 0.4rem;
  padding: 20px;
  width: 300px;
  height: 200px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: left;
  -webkit-box-align: left;
  -ms-flex-align: left;
  align-items: left;
}

.c6 {
  font-style: normal;
  font-weight: bold;
  font-size: 21px;
  line-height: 25px;
  color: #030517;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  top: 0;
}

.c5 {
  font-style: normal;
  font-weight: 800;
  font-size: 28px;
  line-height: 33px;
  color: #FFD836;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  position: relative;
  height: 100px;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  position: absolute;
  bottom: 0;
}

.c10 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  position: absolute;
  bottom: 0;
  right: 0;
}

.c9 {
  font-size: 18px;
  margin-right: 5px;
  color: #FFD836;
}

.c11 {
  font-size: 18px;
  margin-right: 5px;
  color: #FFD836;
}

<div
  class="c0"
  data-testid="WrapperCard"
>
  <div
    class="c1"
    data-testid="motion"
    style="opacity: 0; transform: scale(0.6) translateZ(0);"
  >
    <button
      class="c2"
    >
      <div
        class="c3"
      >
        <div
          class="c4"
        >
          <h2
            class="c5"
          >
            15/11/2021
          </h2>
          <p
            class="c6"
          >
            Titulo da modal
          </p>
        </div>
        <div
          class="c7"
        >
          <h4
            class="c8"
          >
            <svg
              class="c9"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M15 14s1 0 1-1-1-4-5-4-5 3-5 4 1 1 1 1h8zm-7.978-1A.261.261 0 0 1 7 12.996c.001-.264.167-1.03.76-1.72C8.312 10.629 9.282 10 11 10c1.717 0 2.687.63 3.24 1.276.593.69.758 1.457.76 1.72l-.008.002a.274.274 0 0 1-.014.002H7.022zM11 7a2 2 0 1 0 0-4 2 2 0 0 0 0 4zm3-2a3 3 0 1 1-6 0 3 3 0 0 1 6 0zM6.936 9.28a5.88 5.88 0 0 0-1.23-.247A7.35 7.35 0 0 0 5 9c-4 0-5 3-5 4 0 .667.333 1 1 1h4.216A2.238 2.238 0 0 1 5 13c0-1.01.377-2.042 1.09-2.904.243-.294.526-.569.846-.816zM4.92 10A5.493 5.493 0 0 0 4 13H1c0-.26.164-1.03.76-1.724.545-.636 1.492-1.256 3.16-1.275zM1.5 5.5a3 3 0 1 1 6 0 3 3 0 0 1-6 0zm3-2a2 2 0 1 0 0 4 2 2 0 0 0 0-4z"
              />
            </svg>
             
            10
          </h4>
          <h4
            class="c10"
          >
            <svg
              class="c11"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 1024 1024"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372zm47.7-395.2l-25.4-5.9V348.6c38 5.2 61.5 29 65.5 58.2.5 4 3.9 6.9 7.9 6.9h44.9c4.7 0 8.4-4.1 8-8.8-6.1-62.3-57.4-102.3-125.9-109.2V263c0-4.4-3.6-8-8-8h-28.1c-4.4 0-8 3.6-8 8v33c-70.8 6.9-126.2 46-126.2 119 0 67.6 49.8 100.2 102.1 112.7l24.7 6.3v142.7c-44.2-5.9-69-29.5-74.1-61.3-.6-3.8-4-6.6-7.9-6.6H363c-4.7 0-8.4 4-8 8.7 4.5 55 46.2 105.6 135.2 112.1V761c0 4.4 3.6 8 8 8h28.4c4.4 0 8-3.6 8-8.1l-.2-31.7c78.3-6.9 134.3-48.8 134.3-124-.1-69.4-44.2-100.4-109-116.4zm-68.6-16.2c-5.6-1.6-10.3-3.1-15-5-33.8-12.2-49.5-31.9-49.5-57.3 0-36.3 27.5-57 64.5-61.7v124zM534.3 677V543.3c3.1.9 5.9 1.6 8.8 2.2 47.3 14.4 63.2 34.4 63.2 65.1 0 39.1-29.4 62.6-72 66.4z"
              />
            </svg>
            10
          </h4>
        </div>
      </div>
    </button>
  </div>
</div>
`;
